core/server/src/main/java/alluxio/worker/netty/DataServerReadHandler.java

```diff
@@ -0,0 +1,423 @@
+ */
+
+package alluxio.worker.netty;
+
+import alluxio.Configuration;
+import alluxio.Constants;
+import alluxio.PropertyKey;
+import alluxio.network.protocol.RPCMessage;
+import alluxio.network.protocol.RPCProtoMessage;
+import alluxio.network.protocol.databuffer.DataBuffer;
+import alluxio.proto.dataserver.Protocol;
+
+import com.google.common.base.Preconditions;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.locks.ReentrantLock;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * This class handles {@link alluxio.proto.dataserver.Protocol.ReadRequest}s.
+ *
+ * Protocol: Check {@link alluxio.client.block.stream.NettyPacketReader} for more information.
+ * 1. Once a read request is received, the handler creates a {@link PacketReader} which reads
+ *    packets from the block worker and pushes them to the buffer.
+ * 2. The {@link PacketReader} pauses if there are too many packets in flight, and resumes if there
+ *    is room available.
+ * 3. The channel is closed if there is any exception during the packet read/write.
+ */
+@NotThreadSafe
+public abstract class DataServerReadHandler extends ChannelInboundHandlerAdapter {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private static final long PACKET_SIZE =
+      Configuration.getBytes(PropertyKey.WORKER_NETWORK_NETTY_READER_PACKET_SIZE_BYTES);
+  private static final long MAX_PACKETS_IN_FLIGHT =
+      Configuration.getInt(PropertyKey.WORKER_NETWORK_NETTY_READER_BUFFER_SIZE_PACKETS);
+
+  /** The executor to run {@link PacketReader}s.*/
+  private final ExecutorService mPacketReaderExecutor;
+
+  private final ReentrantLock mLock = new ReentrantLock();
+  /** Set to true if the packet reader is active .*/
+  @GuardedBy("mLock")
+  private boolean mPacketReaderActive = false;
+  /**
+   * The next pos to queue to the netty buffer. mPosToQueue - mPosToWrite is the bytes that are
+   * in netty buffer.
+   */
+  @GuardedBy("mLock")
+  private long mPosToQueue = -1;
+  /** The next pos to write to the channel. */
+  @GuardedBy("mLock")
+  private long mPosToWrite = -1;
+
+  /**
+   * This is only updated in the channel event loop thread when a read request starts or cancelled.
+   * No need to be locked.
+   */
+  protected volatile ReadRequestInternal mRequest = null;
+
+  protected abstract class ReadRequestInternal implements Closeable {
+    // This ID can either be block ID or temp UFS file ID.
+    public long mId = -1;
+    public long mStart = -1;
+    public long mEnd = -1;
+    public long mCancelled = Long.MAX_VALUE;
```

Comment:
The position at which this read request is cancelled at. It is used to distinguish a cancel before and after the completion of this read request.  I've updated the comment